generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -0,0 +1,209 @@
+package chipyard.chiptop
+
+import chisel3._
+import chipyard.{Top, System, SystemModule}
+
+//import chisel3.experimental.{DataMirror}
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import freechips.rocketchip.devices.debug.{DebugIO, PSDIO, HasPeripheryDebugModuleImp}
+import freechips.rocketchip.subsystem.{HasExtInterruptsModuleImp, CanHaveMasterAXI4MemPortModuleImp}
+import sifive.blocks.devices.gpio.{HasPeripheryGPIOModuleImp}
+import sifive.blocks.devices.uart.{UARTPortIO, HasPeripheryUARTModuleImp}
+import testchipip.{BlockDeviceIO, CanHavePeripheryBlockDeviceModuleImp, SerialIO, CanHavePeripherySerialModuleImp}
+import icenet.{NICConfig, NICIOvonly, CanHavePeripheryIceNICModuleImp}
+import tracegen.{HasTraceGenTilesModuleImp}
+import chipyard.config.ConfigValName._
+
+case object BuildSystem extends Field[Parameters => SystemModule[System]]((p: Parameters) => Module(LazyModule(new Top()(p)).suggestName("system").module))
+
+abstract class BaseChipTop(implicit val p: Parameters) extends RawModule {
+
+  val systemClock = Wire(Clock())
+  val systemReset = Wire(Reset())
+
+  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+
+  def connectReset(r: Bool): Unit
+
+}
+
+trait HasChipTopSimpleClockAndReset { this: BaseChipTop =>
+  implicit val p: Parameters
+
+  val clock = IO(Input(Clock()))
+  // TODO how to handle async vs sync reset?
+  val reset = IO(Input(Bool()))
+
+  //this.systemClock := DigitalInputIO(clock)
+  //this.systemReset := DigitalInputIO(reset)
+
+  this.systemClock := clock
+  this.systemReset := reset
+
+  def connectReset(r: Bool) { reset := r }
+
+}
+
+trait CanHaveChipTopGPIO { this: BaseChipTop =>
+  implicit val p: Parameters
+  // TODO this becomes analog
+  val gpio: Seq[Bool] = system match {
+    case system: HasPeripheryGPIOModuleImp => {
+      system.gpio.map(_.pins).map(_.map(_.i.ival)).flatten.map { pin =>
+        // TODO sane naming
```

Comment:
Is this supposed to also be the point where the IOCell model is instantiated?

Reply:
Yes